``geth --dev`` test fixture setup

fselmo opened the pull request:
### What was wrong?

Related to Issue #3188 

- Could not build a geth fixture or run tests for latest geth versions using PoS where mining is not available
- Prevented from running tests against latest geth due to some older commands / private chains needing a PoS client as well.

### How was it fixed?

- Running with `geth --dev` for local chain that uses simulated PoS
- Fixture generation with geth running `--dev` mode
- Tests with geth running `--dev` mode

### Todo:
- [x] Clean up the changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://www.shutterstock.com/image-photo/give-me-five-260nw-125450921.jpg)

Commit:
WIP: Generate geth fixture for shanghai with clique consensus

fselmo committed:
Initial squashed commits for getting ``geth --dev`` test fixture working:

- Quicker timeout for unmined wait-for-txn-receipt tests
- Update default fees test for geth --dev mode
- Get rid of all miner start() and stop() references
- Run integration tests with --dev flag
- Test refactoring for ``geth --dev`` test fixture setup
- Remove PoW related endpoint tests; TODO: remove / deprecate endpoints
- Support latest geth versions
- Update geth version in circleCI config + peripherally related changes
- Turn on PoS block identifier tests
- get geth --dev test suite working

reedsa commented on web3/_utils/module_testing/eth_module.py:
```diff
@@ -3078,17 +3073,14 @@ def test_eth_send_transaction(
             "maxFeePerGas": w3.to_wei(4.321, "gwei"),
-            "maxPriorityFeePerGas": w3.to_wei(1, "gwei"),
-            "nonce": w3.eth.get_transaction_count(unlocked_account),
+            "maxPriorityFeePerGas": w3.to_wei(1.2345, "gwei"),
```
Look into this test further.

reedsa commented on web3/_utils/module_testing/web3_module.py:
```diff
@@ -30,6 +31,7 @@
 
 class Web3ModuleTest:
     def test_web3_client_version(self, w3: Web3) -> None:
```
Could add `flaky` to this one.

reedsa commented on tests/integration/generate_fixtures/common.py:
```diff
@@ -223,28 +227,19 @@ def mine_block(w3):
     w3.geth.personal.unlock_account(w3.eth.coinbase, KEYFILE_PW)
-    deploy_txn_hash = factory.constructor().transact({"from": w3.eth.coinbase})
+    deploy_txn_hash = factory.constructor().transact(
+        {"from": w3.eth.coinbase, "gas": 5000000}
```
The docs specify that transactions cost zero gas in `--dev` mode. Unclear why the gas is required to deploy the contract but still digging.

reedsa replied:
Finding that my fixture doesnt generate with this gas value anymore. It might not be necessary with the recent configuration changes and mining.

Commit:
Updates towards ``geth --dev`` test fixture:

- Flaky for replace transaction
- Fix typing
- Loosen assertions on gas
- Nonce value fix
- Loosen asserts for default maxFeePerGas and maxPriorityFeePerGas tests

reedsa commented on tests/integration/test_ethereum_tester.py:
```diff
@@ -319,9 +310,6 @@ class TestEthereumTesterEthModule(EthModuleTest):
     test_eth_fee_history_no_reward_percentiles = not_implemented(
         EthModuleTest.test_eth_fee_history_no_reward_percentiles, MethodUnavailable
     )
-    test_eth_send_transaction_with_nonce = not_implemented(
```
Is this the right approach for these? The tests are passing now but I'd like to know if py-evm was updated or if something else changed?

fselmo committed:
``geth --dev`` test fixture tweaks:

- Add @flaky to tests that expect mining at certain times
- This isn't ideal. Perhaps we can increase the ``dev.period`` (mining
  interval) to make these tests a bit more reliable and hopefully
  the other tests are unaffected.
- Update benchmark to run with ``geth --dev`` setup
- Put back old state of get_logs_without_logs for eth_tester

Commit:
More work towards ``geth --dev`` test fixture:

- Try with --dev.period=5
- Remove assert from test_eth_new_block_filter
- Remove sleep from test_web3_client_version
- Try without flaky
- Fixes test_eth_send_transaction_with_nonce without the need for retries.
- Retries were causing transactions to be sent with the same nonce,
  which in turn requires more gas to go through. Running the test once
  fixes this issue.
- Obtain the transaction count providing the 'pending' argument.
- Fix test_eth_send_transaction_no_max_fee

fselmo commented:
^^ squashed and rebased

Commit:
Additional cleanup for ``geth --dev`` test fixture

pacrob commented on tests/integration/go_ethereum/conftest.py:
```diff
@@ -94,14 +94,17 @@ def base_geth_command_arguments(geth_binary, datadir):
 @pytest.fixture(scope="module")
 def geth_zipfile_version(get_geth_version):
-    if get_geth_version.major == 1 and get_geth_version.minor in [10, 11]:
+    if get_geth_version.major == 1 and get_geth_version.minor in [10, 11, 12, 13]:
```
nit: `py-geth` dropped support for 1.10.x a while ago, probably ok to do here too.

pacrob commented on tests/integration/go_ethereum/test_goethereum_ws.py:
```diff
@@ -48,7 +48,7 @@ def _geth_command_arguments(ws_port, base_geth_command_arguments, geth_version):
             "--allow-insecure-unlock",
         )
-        if geth_version.minor not in [10, 11]:
+        if geth_version.minor not in [10, 11, 12, 13]:
```
Drop 10 here too?

pacrob commented on tests/integration/go_ethereum/test_goethereum_ws_v2/conftest.py:
```diff
@@ -34,7 +34,7 @@ def _geth_command_arguments(ws_port, base_geth_command_arguments, geth_version):
             "--allow-insecure-unlock",
         )
-        if geth_version.minor not in [10, 11]:
+        if geth_version.minor not in [10, 11, 12, 13]:
```
Lots of 10s!

pacrob reviewed (approved):
lgtm! Take or leave geth version comments as you will.

fselmo commented on tests/integration/generate_fixtures/go_ethereum.py:
```diff
@@ -139,22 +137,23 @@ def generate_go_ethereum_fixture(destination_dir):
             genesis_file_path=genesis_file_path,
-            geth_ipc_path=geth_ipc_path,
             geth_port=geth_port,
+            keyfile_pw=keyfile_pw,
```
I thought we got rid of passing this around in `get_geth_process`. We either can leave this (nbd) or, since we know where it lives, we can do what we do in the other instances where we spin up geth: `os.path.join(datadir, "keystore", "pw.txt")`. Take or leave, I just remember removing this at some point.

fselmo commented on tests/integration/go_ethereum/test_goethereum_http.py:
```diff
@@ -57,7 +57,7 @@ def _geth_command_arguments(rpc_port, base_geth_command_arguments, geth_version)
             rpc_port,
             "--http.api",
-            "admin,eth,net,web3,personal,miner,txpool",
+            "admin,eth,net,web3,personal,txpool",
```
pretty sure we can remove `personal` from here as well

reedsa replied:
We still utilize deprecated personal stuff for now so tests require it.

fselmo replied:
Which tests though because we don't need it. `geth --dev` unlocks the coinbase account so we can possibly use that for funding, etc.

fselmo commented on tests/integration/go_ethereum/test_goethereum_ipc.py:
```diff
@@ -25,11 +22,8 @@
     yield from base_geth_command_arguments
     yield from (
-        "--port",
-        geth_port,
```
we can probably remove `--rpc.enabledeprecatedpersonal` from here

reedsa replied:
Also tried without and no luck.

fselmo commented on web3/_utils/module_testing/module_testing_utils.py:
```diff
@@ -49,6 +52,14 @@
     )
 
 
+"""
```
👌🏼

fselmo commented on web3/tools/benchmark/node.py:
```diff
@@ -80,20 +80,23 @@ def _geth_binary(self) -> str:
-            "admin,eth,net,web3,personal,miner",
+            "admin,eth,net,web3",
             "--ipcdisable",
             "--allow-insecure-unlock",
```
I think `--allow-insecure-unlock` can go as well

reedsa replied:
No luck with this either, tests are failing without it.

fselmo replied:
I think the calls we make to unlock account aren't necessary but I could be wrong. We might need to tackle that in a separate PR but we also might be able to use the unlocked accounts that `geth --dev` gives us. I think it's just `coinbase` but it could be others as well, would have to look at the docs.

fselmo replied:
Let's leave it alone for now, this is beyond the scope of the PR I think

Commit:
Address feedback for ``geth --dev`` test fixture

fselmo commented:
lgtm but I can't review it since it's my branch 😆... green light from me though 👍🏼

kclowes reviewed (approved):
LGTM! :shipit:
